lib/backup/azremote/azblob_test.go

```diff
@@ -0,0 +1,190 @@
+			},
+			ExpectedLogs: []string{`Creating AZBlob service client from connection string`},
+		},
+		"base envtemplate package is used and connection string err bubbles": {
+			Dir:           "foo",
+			ExpectedDir:   "foo/",
+			IgnoreFakeEnv: true,
+			Env: map[string]string{
+				envStorageAccCs: "BlobEndpoint=https://test.blob.core.windows.net/;SharedAccessSignature=",
+			},
+			ExpectedErr: "failed to detect any credentials",
+		},
+		"shared key credential err bubbles": {
+			Dir:         "foo",
+			ExpectedDir: "foo/",
+			Env: map[string]string{
+				envStorageAcctName: "test",
+				envStorageAccKey:   "teapot",
+			},
+			ExpectedLogs: []string{`Creating AZBlob service client from account name and key`},
+			ExpectedErr:  `illegal base64 data at input`,
+		},
+		"uses shared key credential": {
+			Dir:         "foo",
+			ExpectedDir: "foo/",
+			Env: map[string]string{
+				envStorageAcctName: "test",
+				envStorageAccKey:   "dGVhcG90Cg==",
+			},
+			ExpectedLogs: []string{`Creating AZBlob service client from account name and key`},
+		},
+		"allows overriding domain name with account name and key": {
+			Dir:         "foo",
+			ExpectedDir: "foo/",
+			Env: map[string]string{
+				envStorageAcctName: "test",
+				envStorageAccKey:   "dGVhcG90Cg==",
+				envStorageDomain:   "foo.bar",
+			},
+			ExpectedLogs: []string{
+				`Creating AZBlob service client from account name and key`,
+				`Overriding default Azure blob domain with "foo.bar"`,
+			},
+		},
+		"can't specify both connection string and shared key": {
+			Dir:         "foo",
+			ExpectedDir: "foo/",
+			Env: map[string]string{
+				envStorageAccCs:    "teapot",
+				envStorageAcctName: "test",
+				envStorageAccKey:   "dGVhcG90Cg==",
+			},
+			ExpectedErr: `only one of connection string, account name and key, or default credential can be specified`,
+		},
+		"just use default is an err": {
+			Dir:         "foo",
+			ExpectedDir: "foo/",
+			Env: map[string]string{
+				envStorageDefault: "true",
+			},
+			ExpectedErr: "failed to detect any credentials",
+		},
+		"uses default credential": {
+			Dir:         "foo",
+			ExpectedDir: "foo/",
+			Env: map[string]string{
+				envStorageDefault:  "true",
+				envStorageAcctName: "test",
+			},
+			ExpectedLogs: []string{`Creating AZBlob service client from default credential`},
+		},
+	}
+
+	for name, test := range cases {
+		t.Run(name, func(t *testing.T) {
+			tlog := &testLogger{}
+
+			logger.SetOutputForTests(tlog)
+			t.Cleanup(logger.ResetOutputForTest)
+
+			fs := &FS{Dir: test.Dir}
+			if test.Env != nil && !test.IgnoreFakeEnv {
+				fs.env = test.Env
+			}
+
+			err := fs.Init()
+			checkErr(t, err, test.ExpectedErr)
+
+			if fs.Dir != test.ExpectedDir {
+				t.Errorf("expected dir %q, got %q", test.ExpectedDir, fs.Dir)
+			}
+
+			tlog.MustContain(t, test.ExpectedLogs...)
+		})
+	}
+}
+
+func checkErr(t *testing.T, err error, shouldContain string) {
```

Comment:
I would suggest to refactor this to avoid the following:
- checking the error mesasge text - replace this with specific errors by type. This will allow to avoid necessity to update tests if the error message will be changed. Seems like errors should be created for the following cases: no credentials, invalid credentials, multiple credentials.
- checking the errors which is returned by underlying Azure library - these tests can start failing after the library updates, so it will require to update tests as well.

Reply:
e8cb18dfd